streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -181,8 +182,16 @@ public StreamTask(final TaskId id,
 
         final TimestampExtractor defaultTimestampExtractor = config.defaultTimestampExtractor();
         final DeserializationExceptionHandler defaultDeserializationExceptionHandler = config.defaultDeserializationExceptionHandler();
+        final Set<String> sourceTopics = topology.sourceTopics();
         for (final TopicPartition partition : partitions) {
-            final SourceNode source = topology.source(partition.topic());
+            final String topicName = partition.topic();
+            if (!sourceTopics.contains(topicName)) {
+                throw new TopologyException(
+                        "Topic not found " + topicName + ". Is the Streams Topology built in a deterministic way?"
```

Comment:
`Topic not found` sounds like as-if the topic was not found in the cluster -- however, what actually happened is that we received a record but the record's topic is unknown in the sub-topology.

Similar to above, "deterministic" is not really easy to understand. I would also not phrase it as a question, but as a statement:
```
... This may happen if different KafkaStreams instances of the same application execute different Topologies. Note that Topologies are only identical if all operators are added in the same order.
```

Or similar.